python/tests/integration/toolbox/test_library_tool.py

```diff
@@ -230,3 +232,92 @@ def iterate_through_version_chain(key):
+    with pytest.raises(SchemaException):
+        lib.read(sym, incomplete=True, date_range=(pd.Timestamp(0), pd.Timestamp(2030, 1, 1)))
+    with pytest.raises(SchemaException):
+        lib.compact_incomplete(sym, append=True, convert_int_to_float=False, via_iteration=False)
+
+    # We change the last append data key to string and verify it's now a string
+    backout_segment = lib_tool.update_append_data_column_type(append_keys[0], "col", str)
+    assert read_append_data_keys_from_ref(sym) == append_keys
+    assert [read_type(key, "col") for key in append_keys] == [str_dtype, str_dtype, DataType.INT64]
+    assert [read_type(key, "other") for key in append_keys] == [DataType.INT64, DataType.INT64, DataType.INT64]
+    assert_frame_equal(lib_tool.read_to_dataframe(append_keys[0]), get_df(1, 5, str))
+
+    # We test that we can backout the change using the returned SegmentInMemory
+    lib_tool.overwrite_segment_in_memory(append_keys[0], backout_segment)
+    assert read_append_data_keys_from_ref(sym) == append_keys
+    assert [read_type(key, "col") for key in append_keys] == [DataType.INT64, str_dtype, DataType.INT64]
+    assert [read_type(key, "other") for key in append_keys] == [DataType.INT64, DataType.INT64, DataType.INT64]
+    assert_frame_equal(lib_tool.read_to_dataframe(append_keys[0]), get_df(1, 5, np.int64))
+
+    # And now make all append data keys ints which makes the symbol readable.
+    lib_tool.update_append_data_column_type(append_keys[1], "col", np.int64)
+    lib_tool.update_append_data_column_type(append_keys[2], "col", np.int64) # This should be idempotent
+    assert read_append_data_keys_from_ref(sym) == append_keys
+    assert [read_type(key, "col") for key in append_keys] == [DataType.INT64, DataType.INT64, DataType.INT64]
+    assert [read_type(key, "other") for key in append_keys] == [DataType.INT64, DataType.INT64, DataType.INT64]
+    entire_symbol = lib.read(sym, incomplete=True, date_range=(pd.Timestamp(0), pd.Timestamp(2030, 1, 1))).data
+    assert_frame_equal(entire_symbol, get_df(6, 0, np.int64))
+
+    # And test that we can overwrite with arbitrary data
+    lib_tool.overwrite_append_data_with_dataframe(append_keys[0], get_df(10, 5, np.int64))
+    assert read_append_data_keys_from_ref(sym) == append_keys
+    assert [read_type(key, "col") for key in append_keys] == [DataType.INT64, DataType.INT64, DataType.INT64]
+    assert [read_type(key, "other") for key in append_keys] == [DataType.INT64, DataType.INT64, DataType.INT64]
+    entire_symbol = lib.read(sym, incomplete=True, date_range=(pd.Timestamp(0), pd.Timestamp(2030, 1, 1))).data
+    assert_frame_equal(entire_symbol, get_df(15, 0, np.int64))
+
+    # And test that compaction now works with the new types
+    lib.compact_incomplete(sym, append=True, convert_int_to_float=False, via_iteration=False)
+    assert read_append_data_keys_from_ref(sym) == []
+    assert_frame_equal(lib.read(sym).data, get_df(15, 0, np.int64))
```

Comment:
Test for appending more incompletes (with the right types) and compacting?

Reply:
Added a test for this in the end of the test.